src/client/Inode.cc

```diff
@@ -137,18 +150,33 @@ void Inode::make_nosnap_relative_path(filepath& p)
 
 void Inode::get_open_ref(int mode)
 {
-  open_by_mode[mode]++;
+  if (snapid != CEPH_NOSNAP)
+    return;
+
+  int bits = (mode << 1) | 1;
+  for (int i = 0; i < CEPH_FILE_MODE_BITS; i++) {
+    if (bits & (1 << i))
+	open_by_mode[i]++;
+  }
   break_deleg(!(mode & CEPH_FILE_MODE_WR));
 }
 
 bool Inode::put_open_ref(int mode)
 {
-  //cout << "open_by_mode[" << mode << "] " << open_by_mode[mode] << " -> " << (open_by_mode[mode]-1) << std::endl;
-  auto& ref = open_by_mode.at(mode);
-  ceph_assert(ref > 0);
-  if (--ref == 0)
-    return true;
-  return false;
+  if (snapid != CEPH_NOSNAP)
```

Comment:
@jtlayton @batrick will this check be an issue since it gets called elsewhere too, not just by https://github.com/ceph/ceph/pull/38964/commits/16b01032b96810618a3b39008e5abdc18df1091e#diff-7a3052fe46aebfed0382c9d0bb9880ea1328add824e0b10c5d551ddfee282cd1R9257 . Similar question for https://github.com/ceph/ceph/pull/38964/commits/16b01032b96810618a3b39008e5abdc18df1091e#diff-314075846a9d6413faad01c659cc8c7d4e626e19c78c3750aef569a0bd812968R153

Reply:
On a  similar note should https://github.com/ceph/ceph/pull/37945/commits/89cc2cda4aa4dd7a53ba3eb8c7b5aa2cb89796fa#diff-314075846a9d6413faad01c659cc8c7d4e626e19c78c3750aef569a0bd812968R152-R154 be removed as well (haven't rebased yet but found it during rebase).

Reply:
I don't know. What's the rationale for skpping refcounting on open snap inodes?

Reply:
It seems the refcounting is done by changing in->snap_cap_refs in the case of snap inodes. See: https://github.com/ceph/ceph/blob/master/src/client/Client.cc#L9301 and the change by Zheng in the same PR: https://github.com/ceph/ceph/pull/38964/files#diff-ec46c4e8098a133d4c998171411583941bb0e9d1a9e570a865ef5c876a9cca94R367. If that's the case, shouldn't there be the same check and snap ref changing here https://github.com/ceph/ceph/blob/master/src/client/Client.cc#L9401 as well?